pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/properties/EnumeratedProperty.java

```diff
@@ -4,97 +4,115 @@
+    implements EnumeratedPropertyDescriptor<E, E> {
 
-    public static final PropertyDescriptorFactory FACTORY = new BasicPropertyDescriptorFactory<EnumeratedProperty>(
-            Enumeration.class) {
+    /** Factory. */
+    public static final PropertyDescriptorFactory<? extends Enumeration> FACTORY // @formatter:off
+        = new SingleValuePropertyDescriptorFactory<Enumeration>(Enumeration.class) { // TODO:cf Enumeration? Object?
+
+            @Override
+            public EnumeratedProperty createWith(Map<PropertyDescriptorField, String> valuesById) {
+                return new EnumeratedProperty<>(nameIn(valuesById),
+                                                descriptionIn(valuesById),
+                                                labelsIn(valuesById),   // this is not implemented
+                                                choicesIn(valuesById),  // ditto
+                                                indexIn(valuesById),    // ditto
+                                                classIn(valuesById),
+                                                0f);
+            }
+        }; // @formatter:on
+
+
+    private final Map<String, E> choicesByLabel;
+    private final Map<E, String> labelsByChoice;
+    private final Class<E> valueType;
 
-        @Override
-        public EnumeratedProperty createWith(Map<String, String> valuesById) {
 
-            return new EnumeratedProperty(nameIn(valuesById), descriptionIn(valuesById), labelsIn(valuesById),
-                    choicesIn(valuesById), indexIn(valuesById), 0f);
-        }
-    };
-
-    /**
-     * Constructor for EnumeratedProperty.
-     *
-     * @param theName
-     *            String
-     * @param theDescription
-     *            String
-     * @param theLabels
-     *            String[]
-     * @param theChoices
-     *            E[]
-     * @param defaultIndex
-     *            int
-     * @param theUIOrder
-     *            float
-     * @throws IllegalArgumentException
-     */
     public EnumeratedProperty(String theName, String theDescription, String[] theLabels, E[] theChoices,
-            int defaultIndex, float theUIOrder) {
-        super(theName, theDescription, theLabels, theChoices, new int[] { defaultIndex }, theUIOrder, false);
+                              int defaultIndex, Class<E> valueType, float theUIOrder) {
+        this(theName, theDescription, CollectionUtil.mapFrom(theLabels, theChoices),
+             theChoices[defaultIndex], valueType, theUIOrder);
     }
 
-    /**
-     * @return Class
-     * @see net.sourceforge.pmd.PropertyDescriptor#type()
-     */
+
+    public EnumeratedProperty(String theName, String theDescription, Map<String, E> labelsToChoices,
+                              E defaultValue, Class<E> valueType, float theUIOrder) {
+        super(theName, theDescription, defaultValue, theUIOrder);
+
+        this.valueType = valueType;
```

Comment:
I don't see the need to support such complex external definitions of enumerates.  We should be able to limit external representations to Strings only for now.  It could always be expanded in the future.

Then we need only focus on internal Java based usages.  These serialize only the value, which I'd expect to use the _label_ String?  So your case for heterogenous _value_ types is good, and should be support.  It's not clear to me though if any usages of `type()` remain for these cases.  If needed, keep current API, otherwise, see if it can be paired down.  Regardless, for posterity, javadoc the design indicating supported/unsupported usages.

If something thorny comes up, it can be discussed further.